src/solvers/solver.py

```diff
@@ -0,0 +1,107 @@
+"""This work was partially funded by the EuroHPC PL project,
+funded in the frame of Smart Growth Operational Programme, topic 4.2.
+
+Authors: Mateusz Hurbol, Justyna Zawalska
+"""
+
+import copy
+from abc import ABC, abstractmethod
+from typing import Dict, List, Tuple, Any
+
+import numpy as np
+from pandas import DataFrame
+
+from src.prepare_problem import (calc_dataframes, extract_tasks,
+                                 find_all_paths_in_dag,
+                                 get_machines_from_wfcommons_file, Task)
+from src.utils.file_management import read_json_file, write_json_file
+
+
+class Solver(ABC):
+    """The base class for solvers."""
+    cost_df: DataFrame
+    runtimes_df: DataFrame
+
+    def __init__(self, wfcommons_filename: str, machines_filename: str, output_filename: str, deadline: float):
+        self.output_filename: str = output_filename
+        self.deadline: float = deadline
+
+        self.wfcommons_data: dict = read_json_file(wfcommons_filename)
+        self.machines: Dict[str, dict] = set_machines(wfcommons_filename, machines_filename)
+        self.paths: List[List[str]] = find_all_paths_in_dag(wfcommons_filename)
+        self.tasks: List[Task] = extract_tasks(self.wfcommons_data)
+        self.cost_df, self.runtimes_df = calc_dataframes(self.machines, self.tasks)
+
+    @abstractmethod
+    def solve(self) -> None:
+        """Finds the solution with a given method and saves it."""
+
+    @abstractmethod
+    def find_solution(self) -> Dict[str, str]:
+        """Uses the selected method for finding the solution."""
+
+    def prepare_cost_function(self, binary_variables: list) -> Tuple[Any, List[Any], List[Any]]:
```

Comment:
```suggestion
    def prepare_cost_function(self, binary_variables: List) -> Tuple[Any, List[Any], List[Any]]:
```
`binary_variables` is list of what?